volatility3/framework/plugins/pgdscan.py

```diff
@@ -0,0 +1,422 @@
+        """
+
+        layer = context.layers[layer_name]
+
+        # check if size is larger than the maxsize limit, and therefore is not saved out.
+        if maxsize <= size:
+            vollog.warning(
+                f"Skip virtual memory dump for {start:#x} as {size} is larger than maxsize limit of {maxsize}"
+            )
+            return None
+
+        file_name = f"pgd.{layer._page_map_offset:#x}.start.{start:#x}.dmp"
+        try:
+            file_handle = open_method(file_name)
+            chunk_size = 1024 * 1024 * 10
+            offset = start
+            while offset < start + size:
+                to_read = min(chunk_size, start + size - offset)
+                data = layer.read(offset, to_read, pad=True)
+                file_handle.write(data)
+                offset += to_read
+        except Exception as excp:
+            vollog.debug(f"Unable to dump virtual memory {file_name}: {excp}")
+            return None
+        return file_handle
+
+    def _merge_mappings_with_gap(self, mappings: List[Tuple[int, int]], gap: int):
+        """
+        Merge overlapping or consecutive ranges based on a specified gap.
+
+        Args:
+            mappings (list of tuples): List of tuples where each tuple is (start, length).
+            gap (int): The gap that determines if two ranges should be merged.
+
+        Returns:
+            list of tuples: The merged mappings, where each tuple is (start, length)
+        """
+
+        # Sort ranges by the start value, mappings should already be in order
+        # but for this to work they MUST be in order so sorting
+        sorted_mappings = sorted(mappings, key=lambda x: x[0])
+
+        merged_mappings = []
+
+        # Start with the first range
+        current_start, current_length = sorted_mappings[0]
+        current_end = current_start + current_length
+
+        for start, length in sorted_mappings[1:]:
+            next_end = start + length
+
+            # If the current range and the next range are within the gap, merge them
+            if start <= current_end + gap:
+                # Extend the current range to include the next one
+                current_end = max(current_end, next_end)
+            else:
+                # If not, add the current range and start a new one
+                merged_mappings.append((current_start, current_end - current_start))
+                current_start, current_length = start, length
+                current_end = next_end
+
+        # Add the last range
+        merged_mappings.append((current_start, current_end - current_start))
+
+        # Return the merged ranges with the gap considered
+        return merged_mappings
+
+    def _generator(self):
+        # get primary layer
+        layer = self.context.layers[self.config["primary"]]
+
+        # Try to move down to the highest physical layer
+        if layer.config.get("memory_layer"):
```

Comment:
We don't yet have a suitable way of guaranteeing this is the lower layer (and this may not work if the lower layer has been swapped out, etc), but until we have something better this is ok.  Be nice to flag it with a FIXME or a TODO, just so we can find it again in the future...

Reply:
Yes - it's something that does pop up a fair bit. I couldn't see an issue tracking it. Do you think it's worthwhile making one? (e.g. so that it's "TODO: Re issue XXXX update to a more suitable way of guaranteeing this is the lower layer")